### PR TITLE
[tests-only] ApiTest. Share space via link

### PR DESCRIPTION
**Pull request by ScharfViktor**
- added api tests.  
- for using some gerkin steps from core I created PR: https://github.com/owncloud/core/pull/40222

### CODE REVIEW

File: tests/acceptance/features/bootstrap/SpacesContext.php [hunk 1]
```diff
@@ -2652,4 +2657,54 @@ public function userHasStoredEtagOfElementOnPathFromSpace($user, $path, $storePa
...
+		$this->featureContext->setResponse(
+			HttpRequestHelper::post(
+				$fullUrl,
+				"",
+				$user,
+				$this->featureContext->getPasswordForUser($user),
+				[],
+				$body
+			)
+		);
```
Comment (SagarGi) [inline tests/acceptance/features/bootstrap/SpacesContext.php:2705]: This function seems to be used in `SpacesContext` repeatedly. May be we can make a single function and use it every time when we use a `POST` method.

File: tests/acceptance/features/apiSpaces/shareSpacesViaLink.feature [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        And the public should be able to download file "/test.txt" from inside the last public link shared folder using the new public WebDAV API with password "123"
+        And the downloaded content should be "some content"
+        Examples:
+            | permissions | permissionsMatching       | password | linkName | expireDate               |
+            | 1           | read                      | 123      | link     | 2042-03-25T23:59:59+0100 |
+            | 5           | read,create               | 123      |          | 2042-03-25T23:59:59+0100 |
+            | 15          | read,update,create,delete |          | link     |                          |
+
+
+    Scenario: An uploader should be abble to upload a file
```
Comment (SagarGi) [inline tests/acceptance/features/apiSpaces/shareSpacesViaLink.feature:50]:
```suggestion
    Scenario: An uploader should be able to upload a file
```

**Review by SagarGi — COMMENTED** [review]
Rest looks good to me :+1:

File: tests/acceptance/features/apiSpaces/shareSpacesViaLink.feature [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            | share_type             | public_link           |
+            | displayname_file_owner | %displayname%         |
+            | displayname_owner      | %displayname%         |
+            | uid_file_owner         | %username%            |
+            | uid_owner              | %username%            |
+            | name                   | <linkName>            |
+        And the public should be able to download file "/test.txt" from inside the last public link shared folder using the new public WebDAV API with password "123"
+        And the downloaded content should be "some content"
+        Examples:
+            | permissions | permissionsMatching       | password | linkName | expireDate               |
```
Comment (kiranparajuli589) [inline tests/acceptance/features/apiSpaces/shareSpacesViaLink.feature:44]:
```suggestion
            | permissions | expectedPermissions       | password | linkName | expireDate               |
```

File: tests/acceptance/features/bootstrap/SpacesContext.php [hunk 4]
```diff
@@ -2652,4 +2657,54 @@ public function userHasStoredEtagOfElementOnPathFromSpace($user, $path, $storePa
 			throw new Exception("Expected stored etag to be some string but found null!");
 		}
 	}
+
+	/**
+	 * @When /^user "([^"]*)" creates public link share of the space "([^"]*)" with settings:$/
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/SpacesContext.php:2662]:
```suggestion
	 * @When /^user "([^"]*)" creates a public link share of the space "([^"]*)" with settings:$/
```

File: tests/acceptance/features/apiSpaces/shareSpacesViaLink.feature [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            | Alice    |
+            | Brian    |
+        And the administrator has given "Alice" the role "Space Admin" using the settings api
+        And user "Alice" has created a space "share space" with the default quota using the GraphApi
+        And user "Alice" has uploaded a file inside space "share space" with content "some content" to "test.txt"
+
+
+    Scenario Outline: A manager can share a space to public via link with different permissions
+        When user "Alice" creates public link share of the space "share space" with settings:
+            | shareType   | 3             |
```
Comment (phil-davis) [inline tests/acceptance/features/apiSpaces/shareSpacesViaLink.feature:21]: This line is not really needed. The step text indicates "public link share" and the step implementation code defaults to "3".

File: tests/acceptance/features/apiSpaces/shareSpacesViaLink.feature [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        Examples:
+            | permissions | permissionsMatching       | password | linkName | expireDate               |
+            | 1           | read                      | 123      | link     | 2042-03-25T23:59:59+0100 |
+            | 5           | read,create               | 123      |          | 2042-03-25T23:59:59+0100 |
+            | 15          | read,update,create,delete |          | link     |                          |
+
+
+    Scenario: An uploader should be abble to upload a file
+        When user "Alice" creates public link share of the space "share space" with settings:
+            | shareType   | 3                        |
```
Comment (phil-davis) [inline tests/acceptance/features/apiSpaces/shareSpacesViaLink.feature:52]: This line is not really needed. The step text indicates "public link share" and the step implementation code defaults to "3".

File: tests/acceptance/features/apiSpaces/shareSpacesViaLink.feature [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            | expireDate  | <expireDate>  |
+        Then the HTTP status code should be "200"
+        And the OCS status code should be "200"
+        And the OCS status message should be "OK"
+        And the fields of the last response to user "Alice" should include
+            | item_type              | folder                |
+            | mimetype               | httpd/unix-directory  |
+            | file_target            | /                     |
+            | path                   | /                     |
+            | permissions            | <permissionsMatching> |
```
Comment (phil-davis) [inline tests/acceptance/features/apiSpaces/shareSpacesViaLink.feature:34]:
```suggestion
            | permissions            | <expectedPermissions> |
```

File: tests/acceptance/features/bootstrap/SpacesContext.php [hunk 8]
```diff
@@ -359,7 +364,7 @@ public function getUserIdByUserName(string $userName): string {
 		}
 		throw new Exception(__METHOD__ . " user with name $userName not found");
 	}
-
+
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/SpacesContext.php:367]:
```suggestion

```

**Comment by ownclouders** [comment]
:boom: Acceptance test [<strong>cs3ApiTests-ocis</strong>](https://drone.owncloud.com/owncloud/ocis/13719/39/1) failed. Further test are cancelled...

**Comment by phil-davis** [comment]
https://drone.owncloud.com/owncloud/ocis/13719/39/6
those cs3-api-tests are flaky - I have seen that a few times recently.

Restarted.

**Comment by micbar** [comment]
It is also a propagation issue, linking to #4251

**Review by SagarGi — COMMENTED** [review]
LGTM :+1:

**Review by micbar — APPROVED** [review]

**Comment by phil-davis** [comment]
https://drone.owncloud.com/owncloud/ocis/13720/41/8
```
runsh: Total unexpected failed scenarios throughout the test run:
apiSpaces/copySpaces.feature:27
apiSpaces/copySpaces.feature:28
```

```
       Failed step: And for user "Brian" the space "Project" should contain these entries:
        response does not contain the entry 'newfolder/insideSpace.txt'
        Failed asserting that an array is not empty.
```

**Comment by ScharfViktor** [comment]
> https://drone.owncloud.com/owncloud/ocis/13720/41/8
> 
> ```
> runsh: Total unexpected failed scenarios throughout the test run:
> apiSpaces/copySpaces.feature:27
> apiSpaces/copySpaces.feature:28
> ```
> 
> ```
>        Failed step: And for user "Brian" the space "Project" should contain these entries:
>         response does not contain the entry 'newfolder/insideSpace.txt'
>         Failed asserting that an array is not empty.
> ```

It's repeating locally. I'll look into it now

**Pull request merged**